### PR TITLE
TASK: Level up to phpstan 2 (Flow 9 adjustments)

### PR DESCRIPTION
**Pull request by mhsdesign**
The upgrade to phpstan level two was introduced via https://github.com/neos/flow-development-collection/pull/3264, this holds the flow 9 specific adjustments.

Related (phpstan level 1) https://github.com/neos/flow-development-collection/pull/3216

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Flow/Classes/Cli/Response.php [hunk 1]
```diff
@@ -47,12 +47,12 @@ class Response
     private $content = '';
 
     /**
-     * @var
+     * @var bool
      */
     private $colorSupport;
```
Comment (kitsunet) [inline Neos.Flow/Classes/Cli/Response.php]: I guess we could also start actually typehinting those? they are private anyways.

File: Neos.Flow/Classes/Core/Booting/Scripts.php [hunk 2]
```diff
@@ -577,8 +578,10 @@ public static function initializeSystemFileMonitor(Bootstrap $bootstrap)
             if (!in_array($packageKey, $packagesWithConfiguredObjects)) {
                 continue;
             }
-            foreach ($package->getAutoloadPaths() as $autoloadPath) {
-                self::monitorDirectoryIfItExists($fileMonitors['Flow_ClassFiles'], $autoloadPath, '\.php$');
+            if ($package instanceof GenericPackage) {
```
Comment (kitsunet) [inline Neos.Flow/Classes/Core/Booting/Scripts.php]: nothing you need to fix here, but we should probably add getAutoloadPaths() to the PackageInterface...

File: Neos.Flow/Classes/Property/PropertyMappingConfiguration.php [hunk 3]
```diff
@@ -361,11 +366,10 @@ public function traverseProperties(array $splittedPropertyPath)
 
         $currentProperty = array_shift($splittedPropertyPath);
         if (!isset($this->subConfigurationForProperty[$currentProperty])) {
-            $type = get_class($this);
             if (isset($this->subConfigurationForProperty[self::PROPERTY_PATH_PLACEHOLDER])) {
                 $this->subConfigurationForProperty[$currentProperty] = clone $this->subConfigurationForProperty[self::PROPERTY_PATH_PLACEHOLDER];
             } else {
-                $this->subConfigurationForProperty[$currentProperty] = new $type;
+                $this->subConfigurationForProperty[$currentProperty] = new static();
```
Comment (kitsunet) [inline Neos.Flow/Classes/Property/PropertyMappingConfiguration.php]: nice one!

File: Neos.Flow/Classes/Cli/Response.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Flow/Classes/Cli/Response.php]: yes we could ... but i have resisted any temptation to use real types to prevent runtime errors or shoot people in the foot via inheritance. @mficzel and me also thought that it would also be the easiest to review ;)

eventually we should migrate of course ^^

File: Neos.Eel/Classes/FlowQuery/OperationInterface.php [hunk 4]
```diff
@@ -42,7 +42,7 @@ public static function isFinal();
      * can work with the $context objects. It can be implemented
      * to implement runtime conditions.
      *
-     * @param array (or array-like object) $context onto which this operation should be applied
+     * @param $context array (or array-like object) $context onto which this operation should be applied
```
Comment (mficzel) [inline Neos.Eel/Classes/FlowQuery/OperationInterface.php]: Should'nt that be `@param array $context`?

Comment (mficzel) [reply Neos.Eel/Classes/FlowQuery/OperationInterface.php]: or `@param array|iterable $context`

Comment (mhsdesign) [reply Neos.Eel/Classes/FlowQuery/OperationInterface.php]: yes good catch ;)

File: Neos.Eel/Classes/FlowQuery/Operations/AbstractOperation.php [hunk 5]
```diff
@@ -80,7 +80,7 @@ public static function getShortName()
     /**
      * {@inheritdoc}
      *
-     * @param array (or array-like object) $context onto which this operation should be applied
+     * @param $context array (or array-like object) $context onto which this operation should be applied
```
Comment (mficzel) [inline Neos.Eel/Classes/FlowQuery/Operations/AbstractOperation.php]: `@param array|iterable $context`

File: Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php [hunk 6]
```diff
@@ -249,14 +249,7 @@ public function loadMetadataForClass($className, ClassMetadata $metadata)
...
                     $uniqueConstraint = ['columns' => $uniqueConstraint->columns];
-                    if (!empty($uniqueConstraint->options)) {
-                        $uniqueConstraint['options'] = $uniqueConstraint->options;
-                    }
-                    if (!empty($uniqueConstraint->name)) {
-                        $primaryTable['uniqueConstraints'][$uniqueConstraint->name] = $uniqueConstraint;
-                    } else {
-                        $primaryTable['uniqueConstraints'][] = $uniqueConstraint;
-                    }
+                    $primaryTable['uniqueConstraints'][] = $uniqueConstraint;
```
Comment (mficzel) [inline Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: Isnt that a change in bahavior as now `name` is not used as array index anymore? Also this would also transfer other keys like `fields` from the constraint. Might well be valid but i would like to unterstand the motivation.

File: Neos.FluidAdaptor/Classes/Core/Parser/Interceptor/ResourceInterceptor.php [hunk 7]
```diff
@@ -91,7 +91,9 @@ public function setDefaultPackageKey($defaultPackageKey)
      */
     public function process(NodeInterface $node, $interceptorPosition, ParsingState $parsingState)
     {
-        /** @var $node TextNode */
+        if (!$node instanceof TextNode) {
+            return $node;
+        }
```
Comment (mficzel) [inline Neos.FluidAdaptor/Classes/Core/Parser/Interceptor/ResourceInterceptor.php:96]: Why is that non breaky? Would unterstand the if-instanceof around the next lines but not before

File: Neos.FluidAdaptor/Classes/View/TemplatePaths.php [hunk 8]
```diff
@@ -356,20 +358,6 @@ protected function sanitizePath($path)
...
-    protected function getPackagePrivateResourcesPath($packageKey)
-    {
-        if (!$this->packageManager->isPackageAvailable($packageKey)) {
-            return null;
-        }
-        $packageResourcesPath = $this->packageManager->getPackage($packageKey)->getResourcesPath();
-
-        return Files::concatenatePaths([$packageResourcesPath, 'Private']);
-    }
-
```
Comment (mficzel) [inline Neos.FluidAdaptor/Classes/View/TemplatePaths.php:375]: Why is this removed specifically? Looks unused but so do other methods in there.

**Review by mficzel — COMMENTED** [review]
Generally like this very much, left some questions

File: Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: The problem was dead code. Phpstan (rightfully) detected that we foreach over `$uniqueConstraint` but in 251 we reassign the variable and thus the check `if (!empty($uniqueConstraint->options)) {` will always fail

the code was introduced as fix via https://github.com/neos/flow-development-collection/commit/012f0e76a3169ed9b30b2dd23698718c4d550782 but it never worked. So removing its seems the right choice? - I have no clue btw about this file ^^

File: Neos.FluidAdaptor/Classes/View/TemplatePaths.php (hunk 8 above)
Comment (mhsdesign) [reply Neos.FluidAdaptor/Classes/View/TemplatePaths.php:375]: yes its unused and phpstan doesnt like it. Other unused methods might be public which is okay.

File: Neos.FluidAdaptor/Classes/Core/Parser/Interceptor/ResourceInterceptor.php (hunk 7 above)
Comment (mhsdesign) [reply Neos.FluidAdaptor/Classes/Core/Parser/Interceptor/ResourceInterceptor.php:96]: The next line always assumes its a TextNode `$node->getText()` but this might fail, as this is a generic processor. Instead when we encounter a node which is not a TextNode we will just early return instead of causing an error.

**Review by mhsdesign — COMMENTED** [review]
.

**Comment by mhsdesign** [comment]
This is now ready to be merged. The tests are now completely green with phpstan level 2.

**NOTE FOR REVIEWERS**
Changes which do not seem self explanatory have been extracted into separate commits with in depth information.
Adding uncritical php type doc comments has been squashed into one commit.

File: Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: Extracted this less simple part into an own pr https://github.com/neos/flow-development-collection/pull/3263

**Review by kitsunet — APPROVED** [review]

**Comment by mhsdesign** [comment]
Thänks @kitsunet ❤️

**Comment by mhsdesign** [comment]
Fyi i also created a backport for Flow 8.3 https://github.com/neos/flow-development-collection/pull/3264

**Review by bwaidelich — APPROVED** [review]
+1 by reading

**Comment by mhsdesign** [comment]
i think it would make sense to merge the backport first to not mess to much with the history: https://github.com/neos/flow-development-collection/pull/3264

The backport basically contains all changes of `TASK: **Pling pling** level up to phpstan 2` as cherry-pick plus an adjustment commit `TASK: Backport phpstan level 2 to Flow 8.3 and ignore to be fixed things` to make things work on 8.3
